src/node/blockstorage.cpp

```diff
@@ -427,6 +427,7 @@ CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
 
 bool IsBlockPruned(const CBlockIndex* pblockindex)
 {
+    LOCK(::cs_main);
```

Comment:
Same as `IsValid()`, except 3 of 4 callers of `IsBlockPruned()` already own `cs_main`. The 4th caller, `blockToJSON()` could be changed to acquire the mutex.

<details>
<summary>suggested change</summary>

```diff
diff --git i/src/node/blockstorage.cpp w/src/node/blockstorage.cpp
index 5074080ed1..2217a4d6c8 100644
--- i/src/node/blockstorage.cpp
+++ w/src/node/blockstorage.cpp
@@ -424,13 +424,13 @@ CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
     }
     return nullptr;
 }
 
 bool IsBlockPruned(const CBlockIndex* pblockindex)
 {
-    LOCK(::cs_main);
+    AssertLockHeld(cs_main);
     return (fHavePruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
 
 // If we're using -prune with -reindex, then delete block files that will be ignored by the
 // reindex.  Since reindexing works by starting at block file 0 and looping until a blockfile
 // is missing, do the same here to delete any later block files after a gap.  Also delete all
diff --git i/src/node/blockstorage.h w/src/node/blockstorage.h
index 7efa7e3b72..0cdab404da 100644
--- i/src/node/blockstorage.h
+++ w/src/node/blockstorage.h
@@ -163,13 +163,13 @@ public:
     {
         Unload();
     }
 };
 
 //! Check whether the block associated with this index entry is pruned or not.
-bool IsBlockPruned(const CBlockIndex* pblockindex);
+bool IsBlockPruned(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
 void CleanupBlockRevFiles();
 
 /** Open a block file (blk?????.dat) */
 FILE* OpenBlockFile(const FlatFilePos& pos, bool fReadOnly = false);
 /** Translation to a filesystem path */
diff --git i/src/rpc/blockchain.cpp w/src/rpc/blockchain.cpp
index 455c32c57a..d7c1572099 100644
--- i/src/rpc/blockchain.cpp
+++ w/src/rpc/blockchain.cpp
@@ -172,13 +172,17 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
             }
             break;
 
         case TxVerbosity::SHOW_DETAILS:
         case TxVerbosity::SHOW_DETAILS_AND_PREVOUT:
             CBlockUndo blockUndo;
-            const bool have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
+            bool have_undo;
+            {
+                LOCK(cs_main);
+                have_undo = !IsBlockPruned(blockindex) && UndoReadFromDisk(blockUndo, blockindex);
+            }
 
             for (size_t i = 0; i < block.vtx.size(); ++i) {
                 const CTransactionRef& tx = block.vtx.at(i);
                 // coinbase transaction (i.e. i == 0) doesn't have undo data
                 const CTxUndo* txundo = (have_undo && i > 0) ? &blockUndo.vtxundo.at(i - 1) : nullptr;
                 UniValue objTx(UniValue::VOBJ);
@@ -925,14 +929,15 @@ static RPCHelpMan getblockheader()
 
     return blockheaderToJSON(tip, pblockindex);
 },
     };
 }
 
-static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
+static CBlock GetBlockChecked(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
+    AssertLockHeld(cs_main);
     CBlock block;
     if (IsBlockPruned(pblockindex)) {
         throw JSONRPCError(RPC_MISC_ERROR, "Block not available (pruned data)");
     }
 
     if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus())) {
@@ -942,14 +947,15 @@ static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
         throw JSONRPCError(RPC_MISC_ERROR, "Block not found on disk");
     }
 
     return block;
 }
 
-static CBlockUndo GetUndoChecked(const CBlockIndex* pblockindex)
+static CBlockUndo GetUndoChecked(const CBlockIndex* pblockindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
+    AssertLockHeld(cs_main);
     CBlockUndo blockUndo;
     if (IsBlockPruned(pblockindex)) {
         throw JSONRPCError(RPC_MISC_ERROR, "Undo data not available (pruned data)");
     }
 
     if (!UndoReadFromDisk(blockUndo, pblockindex)) {
```
</details>

I think it is better to call `UndoReadFromDisk()` also under the same acquisition of `cs_main` as `IsBlockPruned()` (as in the patch above) because that would ensure `nStatus` remains unchanged until after the disk read has completed (i.e. we don't end up with a status that indicates no file on disk and still trying to read the file, if `nStatus` is changed after we read it and before we read the file from disk).

Reply:
I had planned to add lock annotations to `IsBlockPruned()` in a follow-up, but I may as well do it here with the other changes. Done. 

If this pull is now too large to attract re-ACKs, in that case it could be reverted to its original version with the first 3 commits only, and the `nStatus` preparation and change could be proposed separately. We'll see.